README
NAME
    Plack::Middleware::GNUTerryPratchett - Keeping the legacy of Sir Terry
    Pratchett alive forever

VERSION
    version 0.0.1

SYNOPSIS
        use Plack::Builder;
    
        my $app = builder {
            enable "Plack::Middleware::GNUTerryPratchett" ];
            sub {['200', ['Content-Type' => 'text/html'}, ['hello world']]};
        };

DESCRIPTION
    This module adds a "X-Clacks-Overhead" header to every response, to
    celebrate and remember the legacy of Sir Terry Pratchett.

ORIGIN
    (Taken from <http://www.gnuterrypratchett.com>)

         "You know they'll never really die while the Trunk is alive[...]
        It lives while the code is shifted, and they live with it, always Going Home." 
                - Moist von Lipwig, Going Postal, Chapter 13

    In Terry Pratchett's Discworld series, the clacks are a series of
    semaphore towers loosely based on the concept of the telegraph. Invented
    by an artificer named Robert Dearheart, the towers could send messages
    "at the speed of light" using standardized codes. Three of these codes
    are of particular import:

        G: send the message on
        N: do not log the message
        U: turn the message around at the end of the line and send it back again

    When Dearheart died, his name was inserted into the overhead of the
    clacks with a "GNU" in front of it to memorialize him forever (or for at
    least as long as the clacks are standing.)

        "A man is not dead while his name is still spoken." - Going Postal, Chapter 4 prologue

    Keeping the legacy of Sir Terry Pratchett alive forever. For as long as
    his name is still passed along the Clacks, Death can't have him.

SEE ALSO
    <http://www.gnuterrypratchett.com>

AUTHOR
    Yanick Champoux <yanick@cpan.org>

COPYRIGHT AND LICENSE
    This software is copyright (c) 2015 by Yanick Champoux.

    This is free software; you can redistribute it and/or modify it under
    the same terms as the Perl 5 programming language system itself.



Changes
Revision history for Plack-Middleware-GNUTerryPratchett

0.0.1 2015-03-18
 - First version, unleashed on an unsuspecting world.

 [STATISTICS]
 - code churn: 3 files changed, 20 insertions(+), 1 deletion(-)



Makefile.PL

# This file was automatically generated by Dist::Zilla::Plugin::MakeMaker v5.031.
use strict;
use warnings;

use 5.006;

use ExtUtils::MakeMaker;



my %WriteMakefileArgs = (
  "ABSTRACT" => "Keeping the legacy of Sir Terry Pratchett alive forever",
  "AUTHOR" => "Yanick Champoux <yanick\@cpan.org>",
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0
  },
  "DISTNAME" => "Plack-Middleware-GNUTerryPratchett",
  "EXE_FILES" => [],
  "LICENSE" => "perl",
  "MIN_PERL_VERSION" => "5.006",
  "NAME" => "Plack::Middleware::GNUTerryPratchett",
  "PREREQ_PM" => {
    "Plack::Middleware" => 0,
    "Plack::Util" => 0,
    "parent" => 0,
    "strict" => 0,
    "warnings" => 0
  },
  "TEST_REQUIRES" => {
    "File::Spec" => 0,
    "HTTP::Request::Common" => 0,
    "IO::Handle" => 0,
    "IPC::Open3" => 0,
    "Plack::Builder" => 0,
    "Plack::Test" => 0,
    "Test::More" => "0.88"
  },
  "VERSION" => "0.0.1",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


my %FallbackPrereqs = (
  "ExtUtils::MakeMaker" => 0,
  "File::Spec" => 0,
  "HTTP::Request::Common" => 0,
  "IO::Handle" => 0,
  "IPC::Open3" => 0,
  "Plack::Builder" => 0,
  "Plack::Middleware" => 0,
  "Plack::Test" => 0,
  "Plack::Util" => 0,
  "Test::More" => "0.88",
  "parent" => 0,
  "strict" => 0,
  "warnings" => 0
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);





lib/Plack/Middleware/GNUTerryPratchett.pm
package Plack::Middleware::GNUTerryPratchett;
# ABSTRACT: Keeping the legacy of Sir Terry Pratchett alive forever

use strict;
use warnings;

use Plack::Util;

use parent qw( Plack::Middleware );

our $clacks_header = 'X-Clacks-Overhead';
our $value         = 'GNU Terry Pratchett';
 
sub call {
    my $self = shift;

    my $res = $self->app->(@_);

    $self->response_cb(
        $res,
        sub {
            my $res = shift;
            Plack::Util::header_set($res->[1], $clacks_header => $value )
                unless Plack::Util::header_exists($res->[1], $clacks_header );
        });
}

1;

__END__

=head1 SYNOPSIS

    use Plack::Builder;
    
    my $app = builder {
        enable "Plack::Middleware::GNUTerryPratchett" ];
        sub {['200', ['Content-Type' => 'text/html'}, ['hello world']]};
    };

=head1 DESCRIPTION

This module adds a C<X-Clacks-Overhead> header to every response,
to celebrate and remember the legacy of Sir Terry Pratchett.

=head1 ORIGIN

(Taken from L<http://www.gnuterrypratchett.com>)

     "You know they'll never really die while the Trunk is alive[...]
    It lives while the code is shifted, and they live with it, always Going Home." 
            - Moist von Lipwig, Going Postal, Chapter 13 

In Terry Pratchett's Discworld series, the clacks are a series of semaphore towers loosely based on the concept of the telegraph. Invented by an artificer named Robert Dearheart, the towers could send messages "at the speed of light" using standardized codes. Three of these codes are of particular import:

    G: send the message on
    N: do not log the message
    U: turn the message around at the end of the line and send it back again

When Dearheart died, his name was inserted into the overhead of the clacks with a "GNU" in front of it to memorialize him forever (or for at least as long as the clacks are standing.)

    "A man is not dead while his name is still spoken." - Going Postal, Chapter 4 prologue 

Keeping the legacy of Sir Terry Pratchett alive forever.
For as long as his name is still passed along the Clacks, Death can't have him.

=head1 SEE ALSO

L<http://www.gnuterrypratchett.com>
